src/mon/ConnectionTracker.cc

```diff
@@ -203,46 +203,51 @@ void ConnectionTracker::notify_rank_removed(int rank_removed)
   ldout(cct, 20) << "my rank before: " << rank << dendl;
 
   encoding.clear();
-  size_t starting_size = my_reports.current.size();
-  // erase the removed rank from everywhere
-  my_reports.current.erase(rank_removed);
-  my_reports.history.erase(rank_removed);
+  size_t starting_size = peer_reports.size();
+  auto pi = peer_reports.upper_bound(rank_removed);
+  // Remove the target rank and adjust everything that comes after.
   peer_reports.erase(rank_removed);
-  // Move ranks > rank_removed down by 1
-  // First in my_reports' history+current
-  auto ci = my_reports.current.upper_bound(rank_removed);
-  auto hi = my_reports.history.upper_bound(rank_removed);
-  while (ci != my_reports.current.end()) {
-    ceph_assert(ci->first == hi->first);
-    my_reports.current[ci->first - 1] = ci->second;
-    my_reports.history[hi->first - 1] = hi->second;
-    my_reports.current.erase(ci++);
-    my_reports.history.erase(hi++);
+  while (pi != peer_reports.end()) {
+    peer_reports[pi->first - 1] = pi->second; // copy content of next rank to ourself.
+    --peer_reports[pi->first - 1].rank; // update the rank of our new content.
+    peer_reports.erase(pi++); // destroy our next rank and move on.
   }
-  ceph_assert((my_reports.current.size() == starting_size) ||
-	      (my_reports.current.size() + 1 == starting_size));
 
-  // now move ranks down one in peer_reports
-  starting_size = peer_reports.size();
-  auto pi = peer_reports.upper_bound(rank_removed);
-  while (pi != peer_reports.end()) {
-    peer_reports[pi->first - 1] = pi->second;
-    peer_reports.erase(pi++);
+  // Go through everyone, remove & adjust current and history map.
+  for (auto pi = peer_reports.begin(); pi != peer_reports.end(); pi++) {
+    peer_reports[pi->first].current.erase(rank_removed);
+    peer_reports[pi->first].history.erase(rank_removed);
```

Comment:
```pi->second.history.erase(rank_removed)```

etc etc in this loop. Is there a reason you're looking up the peer_report each time?

Reply:
nope, should just use pi->second